codex/sales/reservations.nim

```diff
@@ -217,144 +300,184 @@ proc reserve*(
+    return failure(releaseErr.toErr(ReleaseFailedError))
+
+  reservation.size -= bytes.u256
+
+  # persist partially used Reservation with updated size
+  if err =? (await self.update(reservation)).errorOption:
+
+    # rollback release if an update error encountered
+    trace "rolling back release"
+    if rollbackErr =? (await self.repo.reserve(bytes)).errorOption:
+      rollbackErr.parent = err
+      return failure(rollbackErr)
     return failure(err)
 
-  availability.used = false
-  let r = await self.update(availability)
-  if r.isOk:
-    trace "availability marked unused", id = id.toArray.toHex
-  return r
+  return success()
 
-iterator items*(self: AvailabilityIter): Future[?Availability] =
+iterator items(self: StorableIter): Future[?seq[byte]] =
   while not self.finished:
     yield self.next()
 
-proc availabilities*(
-  self: Reservations): Future[?!AvailabilityIter] {.async.} =
+proc storables(
+  self: Reservations,
+  T: type SomeStorableObject
+): Future[?!StorableIter] {.async.} =
 
-  var iter = AvailabilityIter()
+  var iter = StorableIter()
   let query = Query.init(ReservationsKey)
-
-  without results =? await self.repo.metaDs.query(query), err:
-    return failure(err)
-
-  proc next(): Future[?Availability] {.async.} =
+  when T is Availability:
+    # should indicate key length of 4, but let the .key logic determine it
+    without defaultKey =? AvailabilityId.default.key, error:
+      return failure(error)
+  else:
+    # should indicate key length of 5, but let the .key logic determine it
+    without defaultKey =? key(ReservationId.default, AvailabilityId.default), error:
+      return failure(error)
+
+  without results =? await self.repo.metaDs.query(query), error:
+    return failure(error)
+
+  proc next(): Future[?seq[byte]] {.async.} =
     await idleAsync()
     iter.finished = results.finished
     if not results.finished and
-      r =? (await results.next()) and
-      serialized =? r.data and
-      serialized.len > 0:
+       res =? (await results.next()) and
+       res.data.len > 0 and
+       key =? res.key and
+       key.namespaces.len == defaultKey.namespaces.len:
 
-      return some Json.decode(string.fromBytes(serialized), Availability)
+      return some res.data
 
-    return none Availability
+    return none seq[byte]
 
   iter.next = next
   return success iter
 
-proc unused*(r: Reservations): Future[?!seq[Availability]] {.async.} =
-  var ret: seq[Availability] = @[]
+proc all*(
+  self: Reservations,
+  T: type SomeStorableObject
+): Future[?!seq[T]] {.async.} =
+
+  var ret: seq[T] = @[]
 
-  without availabilities =? (await r.availabilities), err:
-    return failure(err)
+  without storables =? (await self.storables(T)), error:
+    return failure(error)
 
-  for a in availabilities:
-    if availability =? (await a) and not availability.used:
-      ret.add availability
+  # NOTICE: there is a swallowed deserialization error
```

Comment:
Should we log it?

Reply:
Surprisingly, I can't seem to log it because there is an odd compilation error, which I suspect has something to do with questionable. If I change this to:

```nim
for storable in storables.items:
    if bytes =? (await storable):
      without obj =? T.fromJson(bytes), e:
        error "deserialization error", error = e.msg
        continue
      ret.add obj
```
I get the compilation error:
```
/Users/egonat/repos/status-im/nim-codex/tests/codex/sales/testsales.nim(442, 41) template/generic instantiation of `all` from here
/Users/egonat/repos/status-im/nim-codex/codex/sales/reservations.nim(458, 15) template/generic instantiation of `without` from here
/Users/egonat/repos/status-im/nim-codex/vendor/questionable/questionable/withoutresult.nim(36, 3) template/generic instantiation of `without` from here
/Users/egonat/repos/status-im/nim-codex/codex/sales/reservations.nim(459, 15) template/generic instantiation of `error` from here
/Users/egonat/repos/status-im/nim-codex/vendor/nim-chronicles/chronicles.nim(363, 10) template/generic instantiation of `log` from here
/Users/egonat/repos/status-im/nim-codex/vendor/nim-chronicles/chronicles.nim(332, 21) Error: undeclared identifier: 'activeChroniclesStream'
```

This is extrememly odd because there are other uses of `without` and chronicles error logging in the same module and there are no compilation errors there.

Any ideas?

Reply:
For future reference: needed to export chronicles.

Due to the generic type parameter, the context at which this proc is instantiated and called (which is different than at compile time) does not contain chronicles in its scope unless exported by the current module.